### PR TITLE
[Serializer] Improve `AttributeLoader`

### PR DESCRIPTION
**Pull request by dunglas**
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | n/a
| License       | MIT

Improve performance and readability of the attribute loader.

### CODE REVIEW

File: src/Symfony/Component/Serializer/Mapping/Loader/AttributeLoader.php [hunk 1]
```diff
@@ -98,23 +98,26 @@ public function loadClassMetadata(ClassMetadataInterface $classMetadata): bool
...
-                        $attributesMetadata[$property->name]->setIgnore(true);
-                    } elseif ($annotation instanceof Context) {
-                        $this->setAttributeContextsForGroups($annotation, $attributesMetadata[$property->name]);
+
+                        continue;
                     }
 
-                    $loaded = true;
+
+                    match (true) {
```
Comment (stof) [inline src/Symfony/Component/Serializer/Mapping/Loader/AttributeLoader.php:96]: this is broken. When using the annotation reader, the `loadAttributes` method will include *all* annotations of the reflector (unlike attributes which are filtered for known attributes). The existing code is silently ignoring those (there is an implicit no-op `else` case). Your replacement makes it an Error instead (as `match` is required to match a branch), which makes this loader incompatible with classes that use *other* annotations when annotation support is enabled in the serializer component.

Comment (dunglas) [reply src/Symfony/Component/Serializer/Mapping/Loader/AttributeLoader.php:96]: Good catch, thanks! Fixed with `default => null` noop branches.

File: src/Symfony/Component/Serializer/Mapping/Loader/AttributeLoader.php [hunk 2]
```diff
@@ -63,23 +63,12 @@ public function loadClassMetadata(ClassMetadataInterface $classMetadata): bool
...
-                continue;
-            }
-
-            if ($annotation instanceof Context) {
-                $classContextAnnotation = $annotation;
-            }
+            match (true) {
+                $annotation instanceof DiscriminatorMap => $classMetadata->setClassDiscriminatorMapping(new ClassDiscriminatorMapping($annotation->getTypeProperty(), $annotation->getMapping())),
+                $annotation instanceof Groups =>  $classGroups = $annotation->getGroups(),
+                $annotation instanceof Context =>  $classContextAnnotation = $annotation,
```
Comment (alexandre-daubois) [inline src/Symfony/Component/Serializer/Mapping/Loader/AttributeLoader.php]:
```suggestion
                $annotation instanceof Groups => $classGroups = $annotation->getGroups(),
                $annotation instanceof Context => $classContextAnnotation = $annotation,
```

File: src/Symfony/Component/Serializer/Mapping/Loader/AttributeLoader.php [hunk 3]
```diff
@@ -88,33 +77,35 @@ public function loadClassMetadata(ClassMetadataInterface $classMetadata): bool
...
+                    $this->setAttributeContextsForGroups($classContextAnnotation, $attributeMetadata);
                 }
 
                 foreach ($classGroups as $group) {
-                    $attributesMetadata[$property->name]->addGroup($group);
+                    $attributeMetadata->addGroup($group);
                 }
 
                 foreach ($this->loadAttributes($property) as $annotation) {
+                    $loaded = true;
```
Comment (alexandre-daubois) [inline src/Symfony/Component/Serializer/Mapping/Loader/AttributeLoader.php:86]: Is this variable still relevant? It doesn't seem used anywhere in its scope

Comment (dunglas) [reply src/Symfony/Component/Serializer/Mapping/Loader/AttributeLoader.php:86]: Yes, it is returned at the end of the method.

Comment (alexandre-daubois) [reply src/Symfony/Component/Serializer/Mapping/Loader/AttributeLoader.php:86]: Oh right, my bad

**Review by chalasr — APPROVED** [review]
Rather for 7.x I suppose

**Comment by nicolas-grekas** [comment]
Thank you @dunglas.

**Pull request merged**